extension/app/ts/simulation/services/ETHSubscriptionService.ts

```diff
@@ -0,0 +1,92 @@
+import { EthSubscribeParams, JsonRpcMessage, JsonRpcNewHeadsNotification } from '../../utils/wire-types.js'
+import { ErrorWithData } from '../../utils/errors.js'
+import { Future } from '../../utils/future.js'
+
+type Subscription = {
+	callback: (subscriptionId: string, reply: JsonRpcNewHeadsNotification) => void
+	params: EthSubscribeParams,
+	rpcSocket: WebSocket
+}
+
+export class ETHSubscriptionService {
+	private subscriptions = new Map<string, Subscription>()
+	private subscriptionSocket = new Map<WebSocket, string>()
+
+	private webSocketConnectionString: string
+
+	public constructor(webSocketConnectionString: string) {
+		this.webSocketConnectionString = webSocketConnectionString
+	}
+
+	public readonly remoteSubscription = (subscriptionId: string) => {
+		if(this.subscriptions.has(subscriptionId)) {
+			this.subscriptions.get(subscriptionId)?.rpcSocket.close()
+			this.subscriptions.delete(subscriptionId)
+			return true
+		}
+		return false
+	}
+
+	public readonly createSubscription = async (params: EthSubscribeParams, callback: (subscriptionId: string, reply: JsonRpcNewHeadsNotification) => void) => {
+		switch(params.params[0]) {
+			case 'newHeads': {
+				const rpcSocket = new WebSocket(this.webSocketConnectionString)
+				const subscriptionId = new Future<string>()
+
+				rpcSocket.addEventListener('open', _event => {
+					const request = { jsonrpc: '2.0', id: 0, method: 'eth_subscribe', params: ['newHeads'] }
```

Comment:
Out of scope of this PR, but I recommend against using `eth_subscribe` and instead just polling every block.

Reply:
yeah, there's PR for it: https://github.com/DarkFlorist/TheInterceptor/issues/302